File: questions/395. Longest Substring with At Least K Repeating Characters/longest.py
'''
395. Longest Substring with At Least K Repeating Characters
Given a string s and an integer k, return the length of the longest substring of s such that 
the frequency of each character in this substring is greater than or equal to k.

Example 1:

Input: s = "aaabb", k = 3
Output: 3
Explanation: The longest substring is "aaa", as 'a' is repeated 3 times.
Example 2:

Input: s = "ababbc", k = 2
Output: 5
Explanation: The longest substring is "ababb", as 'a' is repeated 2 times and 'b' is repeated 3 times.
'''
class Solution(object):
	def longestSubstring(self, txt, k):
		count = set()
		for c in txt:
			count.add(c)
		n, res= len(count), 0
		for i in range(1, n + 1):
			left, right, no, uni = 0, 0, 0, 0
			cur = [0]*26;
			while right < len(txt):
				if uni <= i:
					idx = ord(txt[right])-97
					#print(idx)
					if cur[idx] == 0:
						uni += 1
					cur[idx] += 1
					if cur[idx] == k:
						no += 1
					right+=1
				else:
					idx = ord(txt[left]) - 97
					if cur[idx] == k:
						no -= 1
					cur[idx] -= 1
					if cur[idx] == 0:
						uni -= 1
					left += 1
				if uni == i and uni == no:
					res = max(res, right - left)
		return res
if __name__ == "__main__":
	l = longest()
	print(l.lon("baaabcb",3))

File: questions/3. Longest Substring Without Repeating Characters/longest.java
/*
Given a string s, find the length of the longest substring without repeating characters.
*/
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] count = new int[256];
        int left = 0, right = 0, res = 0;
        while(right < s.length()){
            char c = s.charAt(right);
            count[(int)c]+=1;
            while(count[(int)c] > 1){
                count[(int)s.charAt(left)]-=1;
                left += 1;
            }
            res = Math.max(res, right - left + 1); 
            right += 1;
        }
        return res;
    }
}